torch/csrc/api/src/nn/modules/embedding.cpp

```diff
@@ -11,27 +11,72 @@
 namespace torch {
 namespace nn {
 
-EmbeddingOptions::EmbeddingOptions(int64_t count, int64_t dimension)
-    : count_(count), dimension_(dimension) {}
+EmbeddingOptions::EmbeddingOptions(int64_t num_embeddings, int64_t embedding_dim) : num_embeddings(num_embeddings), embedding_dim(embedding_dim) {}
 
-EmbeddingImpl::EmbeddingImpl(EmbeddingOptions options) : options(options) {
-  reset();
+EmbeddingImpl::EmbeddingImpl(EmbeddingOptions options): options(options) {
+  EmbeddingImpl::reset();
 }
 
 void EmbeddingImpl::reset() {
-  weight = register_parameter(
-      "weight", torch::empty({options.count_, options.dimension_}));
-  NoGradGuard guard;
-  weight.normal_(0, 1);
+  if (options.padding_idx() != c10::nullopt){
+    if(*options.padding_idx() > 0){
+      assert((*options.padding_idx() < *options.num_embeddings()) && "Padding_idx must be within num_embeddings");
+    }
+    else{
+      assert((*options.padding_idx() >= -(*options.num_embeddings())) && "Padding_idx must be within num_embedding");
+      options.padding_idx(*options.num_embeddings() + *options.padding_idx());
+    }
+  }
+  if (!options._weight().has_value()){
+    torch::nn::init.normal_(weight);
```

Comment:
Before we normalize `weight`, we would need to create a tensor and assign to `weight` first. The Python implementation did this through `self.weight = Parameter(torch.Tensor(num_embeddings, embedding_dim))`, and the equivalent C++ code is `weight = register_parameter("weight", torch::empty({options.num_embeddings(), options.embedding_dim()}));`

Reply:
In the Python implementation, this if-branch contains the following code:
```python
if _weight is None:
    self.weight = Parameter(torch.Tensor(num_embeddings, embedding_dim))
    self.reset_parameters()
```
`self.reset_parameters()` calls the following function:
```python
def reset_parameters(self):
    init.normal_(self.weight)
    if self.padding_idx is not None:
        with torch.no_grad():
            self.weight[self.padding_idx].fill_(0)
```
In the current C++ code, we did `torch::nn::init.normal_(weight);` which matches `init.normal_(self.weight)` in Python `reset_parameters`. However, we still need to translate the remaining of `reset_parameters` to C++:
```python
    ...
    if self.padding_idx is not None:
        with torch.no_grad():
            self.weight[self.padding_idx].fill_(0)
```